Parallel queries

GrantMoyer opened the pull request:
~~Implements IntoParallelIterator for QueryBorrow for rayon integration. I didn't end up using hecs built in BatchedIter, since it doesn't really mesh well with rayon.~~

Implements a ParallelIterator trait on top of bevy_task, which operates on batches. Then implements ParallelIterator for bevy_ecs::query's BatchedIterator.

~~*Still needs to be tested.*~~ Example included

GrantMoyer commented:
I'm reworking this to emulate the std::iter::Iterator API using bevy_tasks.

GrantMoyer commented:
I think this is ready to merge.

cart commented:
looking now

cart commented:
@aclysma  @lachlansneff  no pressure (i'm pretty comfortable reviewing this), but as the authors of bevy_tasks I would welcome your input here.

lachlansneff commented:
This is a very interesting PR. I appreciate how much work you've put into emulating the `ParallelIterator` trait with `bevy_tasks`. My only worry is that it's not quite the same as something like rayon. `.map` won't run that closure on the task pool, so the performance characteristics are not quite the same.

GrantMoyer commented:
Actually, since it sends the whole iterator to the threads in the task pool, the map is run on worker threads. You basically build up a computation description, then send it with batchea of data to worker threads.

lachlansneff commented:
Ah, I see, you have to call `iter_batched` first which gives you an iterator over batches of a particular size. Have you done any benchmarking?

GrantMoyer commented:
No benchmarking yet. I'd like to try it on the 50000 boids example @TheJasonLessard was working on though. I could run a benchmark with a more synthetic test.

lachlansneff commented:
Keep me and @aclysma posted.

aclysma commented:
This looks really neat. I'd love to see some practical uses of it in action. The thing is, I don't find myself using iterator-heavy functional style programming that much as a personal preference. So I'm probably not a great person to give an opinion on this! :)

GrantMoyer commented:
## Benchmarks ##

I've added three very basic benchmarks to the repo. They use the Criterion crate to run the benchmarks with `cargo bench`. The three benchmarks are detailed below.

### Overhead of ParallelIterator ###

Bencmark description: for each element of a 10000 element Vec, do nothing. ParallelIterators use a batch size of 100.

| Test | Average Duration |
|----|---:|
| Iterator | 1.43 ns |
| ParallelIterator 1 Threads | 35,760. ns |
| ParallelIterator 2 Threads | 45,060. ns |
| ParallelIterator 4 Threads | 58,769. ns |
| ParallelIterator 8 Threads | 63,104. ns |
| ParallelIterator 16 Threads | 61,878. ns |
| ParallelIterator 32 Threads | 62,084. ns |

### `for_each()` of expensive operation over a long Vec ###

Benchmark description: for each element of a 10000 element Vec, run a 10000 iteration loop. ParallelIterators use a batch size of 100.

| Test | Average Duration |
|----|---:|
| Iterator | 135.50 ms |
| ParallelIterator 1 Threads | 136.07 ms |
| ParallelIterator 2 Threads | 76.61 ms |
| ParallelIterator 4 Threads | 48.38 ms |
| ParallelIterator 8 Threads | 29.99 ms |
| ParallelIterator 16 Threads | 30.14 ms |
| ParallelIterator 32 Threads | 30.22 ms |


### 10 `map()`s of an expensive operation over a long Vec ###

Benchmark description: For each element of a 10000 element Vec, map a function that doubles its input 1000 times over the Vec 10 times. ParallelIterators use a batch size of 100.

| Test | Average Duration |
|----|---:|
| Iterator | 146.95 ms |
| ParallelIterator 1 Threads | 146.85 ms |
| ParallelIterator 2 Threads | 81.17 ms |
| ParallelIterator 4 Threads | 51.20 ms |
| ParallelIterator 8 Threads | 33.83 ms |
| ParallelIterator 16 Threads | 33.80 ms |
| ParallelIterator 32 Threads | 33.96 ms |

GrantMoyer commented:
I think Criterion may be useful for reproducibly benchmarking code throughout bevy, but I can revert adding the benchmark the repo if we don't want to use Criterion or store benchmarks in the repo in general.

cart commented:
Yeah adding benchmarks is definitely something we've been wanting to do for awhile, and criterion has my preference here. Adding criterion benchmarks to this pr is definitely welcome.

I'll try to wrap up my review today.

cart commented on examples/ecs/parallel_query.rs:
```diff
@@ -0,0 +1,44 @@
+            "Processing entity {}",
+            i.fetch_add(1, atomic::Ordering::Relaxed)
+        );
+        thread::sleep(Duration::from_secs(1));
```
I like that this helps illustrate "expensive" tasks. This example helped prove to me that this works as expected, but I think I would prefer it if the example was a bit more practical.  I don't want newbies thinking that they need to throw thread::sleep or AtomicUsize in their parallel iterators.

Can we make this more like a "minimal example showing how to set up parallel execution"?

cart reviewed (commented):
Fantastic work! I have no comments on the various iterator implementations, they all look correct and close to ideal.

Just a couple of superficial comments on the high level apis then I think this is ready to go.

Eventually I'd love to find a way to make this work (in a way that doesn't tank performance):

```rust
query.par_iter(10).for_each(&pool, |a| a.do_thing())
```

But thats a separate issue that non-parallel queries suffer from too.

GrantMoyer commented:
I renamed `BatchedIter` to `ParIter` and I changed the example to be more practical. The new example is much more beautiful, and also demonstrates `.filter()` nicely.

lachlansneff commented:
@GrantMoyer: Do you think it'd be possible to automatically figure out the optimum batch size based on the number of task pools, so users can just call `.par_iter()` instead of `.par_iter(size_of_batches)`?

aclysma commented:
I would avoid automatic batch sizing. It would be bad to split 32 extremely quick operations across 32 cores, for example. The end-user should make an active decision based on what the workload is like.

aclysma commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -257,3 +272,65 @@ impl<Q: HecsQuery> ChunkIter<Q> {
+}
+
+unsafe impl<'q, 'w, Q: HecsQuery> Send for ParIter<'q, 'w, Q> {}
+unsafe impl<'q, 'w, Q: HecsQuery> Sync for ParIter<'q, 'w, Q> {}
```
not sure about forcing sync for ParIter

GrantMoyer replied:
ParIter is BatchedIter from hecs, but with ParallelIterator implemented instead of Iterator. Hecs implements Sync for BatchedIter, but I haven't thought much about if that's sound.

cart replied:
If we can remove the impls and things still work, lets remove them :smile:

GrantMoyer replied:
Removed.

aclysma commented on crates/bevy_tasks/src/iter/mod.rs:
```diff
@@ -0,0 +1,444 @@
+        (0, None)
+    }
+
+    fn count(mut self, pool: &TaskPool) -> usize {
```
I'm a little concerned people are going to end up putting extremely small units of work into some of these and blast it across threads and end up being a lot slower because of it. "Some people might misuse it" isn't a great reason to not do something but I feel kind of obligated to say it.

GrantMoyer replied:
Maybe I sgould add advice to the docs saying not to use ParallelIterator without profiling your code, because it has significant overhead relative to some tasks.

cart replied:
Yeah a quick doc calling that out is a good idea. Its also worth adding a comment to the example, as many people learn using those.

GrantMoyer replied:
Ideally, I'd like the example to use much more than 128 sprites, since it's probably faster to just do the parallel_query example serially. However, for now, the limiting factor is how long it takes to draw a large number of sprites.

GrantMoyer replied:
I've added documentation in a few places calling out the overhead concerns with ParallelIterator.

aclysma commented on crates/bevy_tasks/src/iter/mod.rs:
```diff
@@ -0,0 +1,444 @@
+    }
+
+    // TODO: Investigate optimizations for less copying
+    fn partition<C, F>(mut self, pool: &TaskPool, f: F) -> (C, C)
```
I definitely understand not wanting to duplicate the docs for the standard library, but a very short description for what it does and a link to the standard docs would be helpful. In fact, I think it's good to say a bit less so that the implementation-specific notes (for example all() not short-circuiting) are more likely to be seen

aclysma replied:
One potential improvement would be to partition by batch instead of item and return an iterator for each set of batches. (I feel like it's something that could be done in a follow-up change although it does change behavior in a non-trivial way.)

GrantMoyer replied:
I've added short decriptions of each ParallelIterator method and linked each to the relevant Iterator docs.

For partition, I'd rather leave performance improvements for the future. I think small breaking changes are still acceptable.

aclysma reviewed (commented):
Took a deeper look, I think the high-level questions I have are:
* Quantity:
    * Do we want something like this at all? (Probably)
    * How much of it should we have? (20% of work that covers 80% of uses, or exhaustive?). I generally favor starting simple/minimal and growing it based on need and practical use than trying to anticipate all possible needs. This is maybe a little more than I personally would start with, but it's a matter of taste and this certainly doesn't seem over the top. :)
* Quality
    * Is the high-level design appropriate as a first-pass? (I think so)
    * Are the implementations correct for what they set out to do?
        * Some of them I'm not sure if they necessarily "work" for parallel iterator
        * A short one-sentence summary, link to in-depth stdlib documentation, and short remarks on implementation/caveats would be helpful without being onerous to write. I wouldn't object to this being done in a follow-on PR.

cart commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -148,6 +149,20 @@ impl<'w, Q: HecsQuery> QueryBorrow<'w, Q> {
+    ///
+    /// Useful for distributing work over a threadpool using the
+    /// ParallelIterator interface.
+    pub fn par_iter<'q>(&'q mut self, batch_size: u32) -> ParIter<'q, 'w, Q> {
```
its worth pointing out that inline sometimes regresses performance. iterators are especially weird in that respect. its worth testing perf for every inline/non-inline decision made.

GrantMoyer replied:
I haven't profiled inline vs not inline here, so I'd prefer to let the compiler make the decision for now.

cart commented on examples/ecs/parallel_query.rs:
```diff
@@ -0,0 +1,64 @@
+    });
+}
+
+// Bounce sprties outside the window
```
nit: sprites

cart commented on examples/ecs/parallel_query.rs:
```diff
@@ -0,0 +1,64 @@
+    for _ in 0..128 {
+        commands
+            .spawn(SpriteComponents {
+                material: materials.add(texture_handle.into()),
```
i want to encourage users to reuse materials when they can (as different materials require gpu state changes during rendering). can we hoist this above the for loop like this?

```rust
let material = materials.add(texture_handle.into());
```

GrantMoyer commented on crates/bevy_tasks/src/iter/adapters.rs:
```diff
@@ -0,0 +1,224 @@
+    pub(crate) iter: Option<P>,
+}
+
+impl<B, P> ParallelIterator<B> for Fuse<P>
```
Fuse just makes sure that once next_batch() returns None, it always returns None.

For flatten, I think it's important to the iter-like API. Flatten necessarily changes the batch size, but note that the batch size isn't always fixed to begin with. In particular, query::ParIter doesn't have a fixed batch size.

GrantMoyer commented on crates/bevy_tasks/src/iter/mod.rs:
```diff
@@ -0,0 +1,444 @@
+
+    // TODO: Investigate optimizations for less copying
+    /// Note that find consumes the whole iterator, and return a Vec of all results
+    fn find<F>(mut self, pool: &TaskPool, f: F) -> Vec<Self::Item>
```
Oops, yeah. It should probably be removed.

GrantMoyer commented on crates/bevy_tasks/src/iter/mod.rs:
```diff
@@ -0,0 +1,444 @@
+                    let mut pos = None;
+                    for item in batch {
+                        if newf(item) {
+                            pos = pos.or(Some(len));
```
It doesn't really male sense to short circuit, since the whole batch has already been consumed.

aclysma replied:
I misread this, didn't catch the .or().

However, do we want to continue calling newf for every element if pos is not None?

GrantMoyer replied:
Good point. I'll remove the unnecessary calls to the predicate.

cart commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -257,3 +272,65 @@ impl<Q: HecsQuery> ChunkIter<Q> {
+}
+
+unsafe impl<'q, Q: HecsQuery> Send for Batch<'q, Q> {}
+unsafe impl<'q, Q: HecsQuery> Sync for Batch<'q, Q> {}
```
Yeah lets remove these if we can.

GrantMoyer replied:
Turns out only `unsafe impl Send for Batch {}` is needed.

GrantMoyer commented:
I rebased on master to avoid a conflict with the recent ParallelExecutor improvement.

cart commented:
CI failure was from a nightly clippy update that added new criteria. I've resolved the issues in master.

cart commented:
Ok I think this is good to merge. I do share @aclysma's concern that we may have started with a "bigger" implementation than we should have, but I'm happy enough with each iterator impl that I'm comfortable merging this as-is.

Thanks for your hard work @GrantMoyer!
